### PR TITLE
[BE] PR 1/3: Leaderboard Foundation - DTOs and Repository Layer (#1122)

### PR DESCRIPTION
**Pull request by Lucy-SD**
## Chore: Leaderboard Foundation: DTOs and Repository Layer

**Closes sub-issue:**  https://github.com/orgs/IT-Academy-BCN/projects/26/views/2?pane=issue&itemId=163754080&issue=IT-Academy-BCN%7Cita-challenges-backend%7C1122

###  📝 Description
Create the foundation for the leaderboard feature by implementing the DTOs that will be exposed to users and the repository layer with MongoDB aggregation queries. This task establishes the data contract and database access layer that subsequent tasks will build upon.

### ✅ Acceptance Criteria
- [x] `LeaderboardEntryDto`  created with `username` (String) and `totalPoints` (int) fields.
- [x] `LeaderboardResponseDto`  created with `leaderboard` (List).
- [x] DTOs placed in `com.itachallenge.challenge.dto.gamification` package with proper Lombok annotations.
- [x] `UserScoreRepository ` updated with aggregation methods  `aggregateUserScores` () returning `Flux<LeaderboardAggregationResult>`.
- [x] `LeaderboardAggregationResult` helper class created in `com.itachallenge.gamification.repository.project` to map MongoDB aggregation results (includes `username`, `totalPoints`).

### 🔧 Technical Details
- **DTO Package**: `com.itachallenge.challenge.dto.gamification`.
- **Repository Package**: `com.itachallenge.gamification.repository`.
- **Aggregation Helper**: `com.itachallenge.gamification.repository.project`.
- **MongoDB Aggregation Pipeline**:
  ```
  [
            "{$sort: {user_id: 1, created_at: -1}}",
            """
             {
               $group: {
                 _id: '$user_id',
                 username: {$first: '$username'},
                 totalPoints: {$sum: '$points_earned'}
               }
             }
            """,
            "{$project: {_id: 0, username: 1, totalPoints: 1}}",
            "{$sort: {totalPoints: -1}}"
  ]

### 💰 Value Provided
Establishes the database query capability and API contract for the leaderboard feature. The repository can now efficiently aggregate user scores.

### 🔓 Unlocks
PR 2/3: https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/1131 Service layer implementation using repository methods.

### 🔗 Dependencies
* None - This is the foundation task.

### 🧪 Testing Performed
* **Unit Tests:**
✅ `LeaderboardEntryDtoTest`: JSON serialization/deserialization, builder pattern, equals/hashCode.
✅ `LeaderboardResponseDtoTest`:  JSON structure validation, pagination fields, empty list handling.

* **Integration Tests:**
✅ `UserScoreRepositoryIntegrationTest with Testcontainers:
       - Verify correct aggregation of points per user.
       - Verify last username update is shown.
       - Test empty DB case.

### 🔄 Definition of Done (DoD)
* All DTOs created and reviewed.
* Repository methods implemented and reviewed.
* All tests passing in CI pipeline.
* Code coverage >80% for new code.
* No SonarLint issues.
* CHANGELOG.md updated.

### 👨‍💻 Technical Debts
* Using `$first` for `username` in aggregation assumes the fist username in chronological order is correct. If usernames can be updated, this might show outdated names. Deferred to future sprint if needed.
* No caching implemented yet - will be added if performance becomes an issue.
* The aggregation includes userId internally but it's not exposed in the final DTO - this is by design for future extensibility.

### CODE REVIEW

**Review by carlosPc1987 — APPROVED** [review]
Tests are clear and cover the important cases: aggregation per user, most recent username when it changes, correct sum of points, and empty DB.
Good use of Testcontainers.

**Review by Andrescalvo22 — APPROVED** [review]
Approved :)

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java
Comment (JungleGiu) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java:1]: I'm thinking if this is a DTO why should it have a setter?
Since this is a read-only response DTO, @Setter adds mutability that nothing should need. @Builder + @AllArgsConstructor already covers construction. Worth dropping to make the contract clearer.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDto.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.itachallenge.challenge.dto.gamification;
+
+import lombok.*;
+
+import java.util.List;
+
+@Getter
+@Setter
```
Comment (JungleGiu) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDto.java]: same here

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java
Comment (Lucy-SD) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java:1]: Good observation. Updated classes and tests = )

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.*;
+import lombok.extern.jackson.Jacksonized;
+
+@Getter
+@Builder
+@Jacksonized
+@AllArgsConstructor
+public class LeaderboardEntryDto {
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java:11]: The name `LeaderboardEntryDto` can be confusing — `Entry` typically suggests an input rather than an output element. Consider whether the name accurately reflects what this object represents in the context of the response. What would be a more descriptive name that clearly communicates its purpose?

File: itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java [hunk 3]
```diff
@@ -9,4 +10,20 @@
 public interface UserScoreRepository extends ReactiveMongoRepository<UserScoreDocument, UUID> {
 
     Flux<UserScoreDocument> findByUserIdOrderByCreatedAtDesc(UUID userId);
+
+    @Aggregation(pipeline = {
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java:15]: Before merging these elements, take a look at `UserScoreServiceImpl`. How is `totalPoints` calculated there? Now compare it to the method used in `aggregateUserScores()`. Do you notice any inconsistencies? What are the implications of having two different ways of calculating the same value in the same project? How should we proceed in these cases?

UserScoreAggregation is placed in the repository package, but it represents a data mapping object rather than a repository. Does this placement accurately reflect the responsibility of this class? Where else in the project structure could it fit better?

File: itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreAggregation.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.itachallenge.gamification.repository;
+
+import lombok.*;
+
+@Getter
+@Builder
+@AllArgsConstructor
+public class UserScoreAggregation {
```
Comment (ivilarop) [inline itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreAggregation.java]: UserScoreAggregation, LeaderboardEntryDto and LeaderboardResponseDto are all specific to the leaderboard use case. Is the naming of UserScoreAggregation consistent with the other classes in this feature? What would make the naming more coherent across the leaderboard-specific classes?

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Test
+    void givenValidLeaderboardEntry_whenSerialize_thenCorrectJsonStructure() {
+        LeaderboardEntryDto dto = LeaderboardEntryDto.builder()
+                .username(TEST_USERNAME)
+                .totalPoints(TEST_POINTS)
+                .build();
+
+        JsonContent<LeaderboardEntryDto> result;
+        try {
```
Comment (ivilarop) [inline itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java]: JUnit allows you to declare `throws IOException` directly in the method. Manual try/catch blocks are unnecessary and add noise.

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                .hasJsonPathNumberValue("@.total_points", TEST_POINTS);
+    }
+
+    @Test
+    void givenValidJson_whenDeserialize_thenCorrectLeaderboardEntry() {
+        String jsonContent = String.format("{\"username\":\"%s\",\"total_points\":%d}",
+                TEST_USERNAME, TEST_POINTS);
+
+        LeaderboardEntryDto dto;
+        try {
```
Comment (ivilarop) [inline itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java]: JUnit allows you to declare `throws IOException` directly in the method. Manual try/catch blocks are unnecessary and add noise.

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.boot.test.json.JacksonTester;
+import org.springframework.boot.test.json.JsonContent;
+
+import java.io.IOException;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.fail;
+
+@JsonTest
+class LeaderboardEntryDtoTest {
```
Comment (ivilarop) [inline itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java:14]: Looking at `LeaderboardEntryDtoTest`, there are a few things to consider:

1. **Unnecessary try/catch blocks**: JUnit 5 allows declaring `throws IOException` directly on the test method. The manual try/catch adds noise without adding value.

2. **Redundant tests**: `givenBuilderPattern_whenCreateDto` and `givenAllArgsConstructor_whenCreateDto` both verify the same thing — that fields are set correctly. Lombok's `@Builder` and `@AllArgsConstructor` are already tested by Lombok itself, so these tests don't add real coverage.

3. **`givenTwoDtosWithSameValues_whenCompare`** only compares fields individually and checks they are not the same instance — it doesn't actually test `equals/hashCode`. If that's the intention, use `assertThat(dto1).isEqualTo(dto2)` instead.

4. **`givenZeroPoints`** doesn't add meaningful coverage beyond what the serialization test already covers.

The tests that truly matter for this DTO are:
- Serialization: verifies `@JsonProperty("total_points")` produces the correct snake_case JSON
- Deserialization: verifies the incoming JSON maps correctly to the DTO fields

Could you simplify the test class to focus on these two? If the additional tests were added to meet Sonar coverage requirements, please clarify — but the goal should be meaningful coverage, not just hitting a number.

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDtoTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import org.springframework.boot.test.json.JsonContent;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.fail;
+
+@JsonTest
+class LeaderboardResponseDtoTest {
```
Comment (ivilarop) [inline itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDtoTest.java:15]: Same observations as in `LeaderboardEntryDtoTest`:

1. **Unnecessary try/catch blocks**: Use `throws IOException` directly on the test method instead.

2. **Redundant tests**: `givenBuilderPattern_whenCreateDto` and `givenAllArgsConstructor_whenCreateDto` only verify that Lombok works — they don't add meaningful coverage.

The tests that truly matter for this DTO are:
- Serialization with a list of entries: verifies the JSON structure is correct
- Deserialization: verifies the JSON maps correctly to the DTO
- Empty list: valid edge case worth keeping

3. **Empty list assertion is incomplete**: 
```java
// Current - only checks the array exists
assertThat(result).hasJsonPathArrayValue("@.leaderboard");

// Should also verify it's empty
assertThat(result).hasEmptyJsonPathValue("@.leaderboard");
```

Could you simplify the test class to focus on these three meaningful cases? If the additional tests were added to meet Sonar coverage requirements, please clarify — but the goal should be meaningful coverage, not just hitting a number.

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreAggregationTest.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package com.itachallenge.gamification.repository;
+
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class UserScoreAggregationTest {
```
Comment (ivilarop) [inline itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreAggregationTest.java]: Both tests in `UserScoreAggregationTest` only verify that Lombok's `@AllArgsConstructor` and `@Builder` are working correctly; this is Lombok's responsibility, not ours. `UserScoreAggregation` is a simple mapping object with no business logic, so there's nothing relevant to test here.

Were these tests added solely to meet Sonar coverage requirements? Have you tried removing these tests and seeing if Sonar fails?

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreRepositoryIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+import org.testcontainers.junit.jupiter.Testcontainers;
+import reactor.core.publisher.Flux;
+import reactor.test.StepVerifier;
+
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+@DataMongoTest
+@Testcontainers(disabledWithoutDocker = true)
+class UserScoreRepositoryIntegrationTest {
```
Comment (ivilarop) [inline itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreRepositoryIntegrationTest.java:23]: The integration test setup with Testcontainers is correct and the cases covered are relevant. However, there are a few things to address:

1. **Duplicated tests**: `givenMultipleUsersScores_whenAggregateUserScores` and `givenUserWithMultipleScores_whenAggregateUserScores` verify exactly the same thing — username + totalPoints for all 3 users. One of them should be removed.

2. **`givenUserWithMultipleUsernames`** only verifies the username without the points, which is already covered by the sorting test. This doesn't add meaningful coverage on its own.

3. **`deleteAll()` is called twice in setUp()**:
```java
// In setUp()
userScoreRepository.deleteAll().block(); // first call
...
private void createTestData() {
    userScoreRepository.deleteAll().block(); // second call — unnecessary
}
```
Pick one place to clean up and remove the duplicate.

4. **Unnecessary null checks**: With `@DataMongoTest`, Spring always injects the repository. `if (userScoreRepository != null)` will never be false and adds noise.

The three tests that truly matter here are:
- Correct descending sort + sum of points per user
- Most recent username is returned when a user has multiple usernames
- Empty database returns no results

Could you simplify the test class to focus on these three cases?

File: CHANGELOG.md [hunk 11]
```diff
@@ -3,6 +3,17 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased] - 2026-03-11
```
Comment (ivilarop) [inline CHANGELOG.md:15]: Remember to modify it when you have made all the changes

**Review by ivilarop — CHANGES_REQUESTED** [review]

**Review by JungleGiu — APPROVED** [review]

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java (hunk 5 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java]: updated.

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java (hunk 6 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java]: updated.

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDtoTest.java (hunk 8 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDtoTest.java:15]: Thanks for the observations.

 1. all try-catch blocks have been deleted.

2. both redundant tests deleted.

3. added empty verification.

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java (hunk 7 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java:14]: Thanks for the feedback.

1. All try-catch bolcks deleted.

2., 3. y 4. All reduntant tests has been deleted.

We are now focusing in serialization adn deserialization only.

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreAggregationTest.java (hunk 9 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreAggregationTest.java]: This was originaly implemented for Sonar reasons only. Deleted.

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreRepositoryIntegrationTest.java (hunk 10 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreRepositoryIntegrationTest.java:23]: Duplication was eliminated throughout the entire class, and repo !=null was also deleted. Now we only focus in the 3 cases you mentioned.

I've also rearranged the private method position to improve readability and ease of understanding.

File: CHANGELOG.md (hunk 11 above)
Comment (Lucy-SD) [reply CHANGELOG.md:15]: updated with changes applied = )

**Review by Lucy-SD — COMMENTED** [review]
Thanks @ivilarop, for all your feedback.

Many tests were present because Sonar made things complicated for me at first, but I should have done a proper cleanup once I found the solution.

Regarding naming, class placement, and the aggregation pipeline, I'd appreciate a heads-up on how we want to align as a team. I've included my reasons for the advantages and disadvantages of each issue.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java (hunk 2 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java:11]: Analyzing the naming within the overall project context, we can see that Gamification already has a PointHistoryEntryDto that follows the exact same pattern: it represents an item within a collection (point history). Keeping LeaderboardEntryDto provides semantic consistency with the existing structure and clearly communicates that it's an individual item within the leaderboard list. Alternatives could be LeaderboardUserDto or LeaderboardItemDto, and although I believe they would break this consistency without offering a clear benefit, I'm willing to rename it if we consider it the best strategy.

File: itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java (hunk 3 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java:15]: We have 2 things to analyze in this comment:

---

1. You've correctly identified a difference in the calculation methods. However, after analyzing the business context, we see that:

* getUserPointsHistory: by design, it needs to retrieve individual documents to build the history timeline. Once we have those documents, summing their points in Java is efficient and avoids an additional database query.

* aggregateUserScores: only requires the total points per user. Performing the aggregation directly in MongoDB is optimal in terms of performance and scalability, especially considering that the leaderboard may have many users.

Unifying the methods would mean sacrificing performance in one of the two cases. I believe it's acceptable to maintain both, as they address different needs. If, in the future, the history also only requires totals, we could refactor, but for now, this decision prioritizes efficiency. Let me know if we prefer to choose another solution.

I'm also preparing a technical debt report: if users database grows to huge amount of active users with frequent score events, we should consider adding a pre-calculated `total_scores` field to user's database.

---

2. Regarding the location of UserScoreAggregation (or its renamed version), I've analyzed the project structure and see that while we use mapper in submissions and helper in challenges, there's no precedent for a projection/aggregation class like this one.
To keep responsibilities and structures clean, I propose creating the repository.projection subpackage within gamification to house UserScoreAggregation (or its renamed version). With this we:

* keep the class close to the repository that uses it (high cohesion).

* follow Spring Data's pattern where projections are usually in subpackages.

* avoids having loose classes in the repository package.

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java
Comment (JungleGiu) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java:1]: Good implementation!

**Review by JungleGiu — APPROVED** [review]

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDto.java (hunk 1 above)
Comment (JungleGiu) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDto.java]: 👍🏻

**Review by JungleGiu — APPROVED** [review]

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java (hunk 2 above)
Comment (carlosPc1987) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java:11]: I agree with Lucy’s defense: LeaderboardEntryDto is a reasonable and consistent name within the existing Gamification module.

Internal consistency: we already have PointHistoryEntryDto, following the same pattern where Entry represents an element within a collection. Keeping LeaderboardEntryDto reinforces that coherence and makes the model more predictable for anyone consuming or reading the API.

Contextual semantics: in a leaderboard response, entry is naturally understood as a row or record in the ranking (user + score/position, etc.). It doesn’t imply “input”; in APIs and DTOs it’s commonly used to mean a record within a list.

Alternatives don’t clearly improve things:
- LeaderboardUserDto could suggest it represents a user profile rather than a ranking row.
- LeaderboardItemDto is more generic and less meaningful than Entry.
- 
If the goal is to eliminate any possible ambiguity, LeaderboardRowDto would be a very explicit option, but I don’t think it’s necessary if we prioritize consistency with the rest of the project.

File: itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java (hunk 3 above)
Comment (carlosPc1987) [reply itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java:15]: +1 to what Lucy mentioned. These aren’t simply “two different ways of doing the same calculation,” but rather two distinct use cases: the history endpoint already needs to fetch the underlying events, and summing them in Java avoids an extra query, whereas the leaderboard only requires per‑user totals, and aggregating directly in MongoDB is the most efficient and scalable approach.
I also agree with placing the projection under gamification.repository.projection, as it better reflects its responsibility and aligns with Spring Data conventions.

**Review by carlosPc1987 — APPROVED** [review]
nice work Lucy!

**Comment by Lucy-SD** [comment]
## All Tests Still Passing After Updates = )

image

**Review by Andrescalvo22 — APPROVED** [review]
✅ All requested changes have been properly addressed.
The code and tests have been improved in terms of clarity, consistency, and structure.

Checklist:

✔ Improved naming consistency across leaderboard-related classes (LeaderboardAggregationResult aligns well with the feature context)

✔ Proper restructuring of aggregation class into repository.projection package (better responsibility separation)

✔ Removal of redundant and low-value tests in DTO test class

✔ Focus on meaningful test coverage (serialization and deserialization)

✔ Elimination of duplicated integration tests

✔ Removal of unnecessary null checks

✔ Test suite now focuses on key scenarios:

correct sorting and total points aggregation

handling of most recent username

empty database behavior

Overall, the implementation is now cleaner, more consistent, and aligned with best practices.

File: itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreAggregation.java (hunk 4 above)
Comment (Lucy-SD) [reply itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreAggregation.java]: You're absolutely right on this point; I could not think of a coherent and cohesive name at first, but after analyzing it several times, I believe `LeaderboardAggregationResult` is the best option to unify this trio of actions and clarify each's functionality. What do you think ¿?

File: itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java (hunk 2 above)
Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDto.java:11]: Fair point — the consistency argument with `PointHistoryEntryDto` is solid, and `Entry` is semantically correct in a leaderboard context (a row in a ranking, not an input). Happy to keep `LeaderboardEntryDto`. Resolving this.

File: itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java (hunk 3 above)
Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java:15]: The technical decision is sound and well-justified: using `@Aggregation` for the leaderboard and keeping the projection classes in `repository/projection` makes sense.

However, this pull request introduces two new patterns that didn't exist in the project:
- `@Aggregation` pipelines in Spring Data MongoDB
- `repository/projection` subpackage for projection classes

This is an important principle to keep in mind: **whenever a new way of working is introduced, it must be documented**. Without documentation, the next developer won't know the pattern exists and will reinvent it differently or place it inconsistently, creating precisely the kind of technical debt that is difficult to resolve later.

Please document these patterns in an accessible location, such as the changelog or the README file for the gamification module, so that the rest of the team can consistently follow them in future tasks.

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java (hunk 7 above)
Comment (ivilarop) [reply itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardEntryDtoTest.java:14]: ok well done. close conversation

File: itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDtoTest.java (hunk 8 above)
Comment (ivilarop) [reply itachallenge-challenge/src/test/java/com/itachallenge/challenge/dto/gamification/LeaderboardResponseDtoTest.java:15]: ok well done. close conversation

File: CHANGELOG.md (hunk 11 above)
Comment (ivilarop) [reply CHANGELOG.md:15]: ok well done. close conversation

File: itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreRepositoryIntegrationTest.java (hunk 10 above)
Comment (ivilarop) [reply itachallenge-challenge/src/test/java/com/itachallenge/gamification/repository/UserScoreRepositoryIntegrationTest.java:23]: ok well done. close conversation

**Review by ivilarop — CHANGES_REQUESTED** [review]

File: itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java (hunk 3 above)
Comment (ivilarop) [reply itachallenge-challenge/src/main/java/com/itachallenge/gamification/repository/UserScoreRepository.java:15]: I know you're working on the README for the gamification module — great initiative! No need to rush it. Finish it at your own pace and open a separate PR for it. I'm going to merge the 3 PRs now so we don't block the sprint. Thanks for the effort!

**Review by ivilarop — APPROVED** [review]
